app/services/promo/registration_stats_report_generator.rb

```diff
@@ -2,113 +2,137 @@
+    @is_geo = is_geo
   end
 
   def perform
-    # Fetch the most recent stats
-    promo_registrations = PromoRegistration.where(referral_code: @referral_codes)
-    Promo::RegistrationsStatsFetcher.new(promo_registrations: promo_registrations).perform
-    promo_registrations.reload
-
-    # Build the report contents
     report_contents = {}
-    promo_registrations.each do |promo_registration|
-      # Pull all statistics associated with a code
-      events = JSON.parse(promo_registration.stats)
-
-      # Select only those within date range supplied
-      events_within_report_period = events.select { |event|
-        (event["ymd"].to_date >= @start_date) && (event["ymd"].to_date <= @end_date)
-      }
-
-      if @reporting_interval == PromoRegistration::RUNNING_TOTAL
-        base_case = {PromoRegistration::RETRIEVALS => 0, PromoRegistration::FIRST_RUNS => 0, PromoRegistration::FINALIZED => 0 }
-        running_total_report_contents = events_within_report_period.reduce(base_case) { |aggregate_stats, event|
-          aggregate_stats[PromoRegistration::RETRIEVALS] += event[PromoRegistration::RETRIEVALS]
-          aggregate_stats[PromoRegistration::FIRST_RUNS] += event[PromoRegistration::FIRST_RUNS]
-          aggregate_stats[PromoRegistration::FINALIZED] += event[PromoRegistration::FINALIZED]
-          aggregate_stats.slice(PromoRegistration::RETRIEVALS, PromoRegistration::FIRST_RUNS, PromoRegistration::FINALIZED)
-        }
-
-        report_contents_for_referral_code = {
-          @start_date => running_total_report_contents
-        }
-
-        report_contents["#{promo_registration.referral_code}"] = report_contents_for_referral_code
-      else
-        report_contents_for_referral_code = empty_report_contents_for_referral_code(@start_date, @end_date, @reporting_interval)
-        events_within_report_period.each do |event|
-          interval_start_date = coerce_date_to_start_or_end_of_reporting_interval(event["ymd"].to_date, @reporting_interval, true)
-          existing_event = report_contents_for_referral_code[interval_start_date]
-          existing_event[PromoRegistration::RETRIEVALS] += event[PromoRegistration::RETRIEVALS]
-          existing_event[PromoRegistration::FIRST_RUNS] += event[PromoRegistration::FIRST_RUNS]
-          existing_event[PromoRegistration::FINALIZED] += event[PromoRegistration::FINALIZED]
+
+    events = fetch_stats
+    events = select_events_within_report_range(events)
+    events = fill_in_events_with_no_activity(@referral_codes, events)
+    events_by_referral_codes = group_events_by_referral_code(events)
+
+    events_by_referral_codes.each do |referral_code, events|
+      events = order_events_by_date(events)
+      events_by_referral_code_for_interval = group_events_by_date(events)
+
+      # Sum the events per interval, or per interval per country if @is_geo
+      per_interval_totals = {}
+      events_by_referral_code_for_interval.each do |date, events|
```

Comment:
IMO the value should be `events_on_date` or something similar.

Reply:
And you could rename the function and save a line/variable declaration like so:

```
events_by_referral_codes.each do |referral_code, events|
  group_events_by_date(order_events_by_date(events)).each do |date, events_on_date|
    ...
```

@corymcdonald @nvonpentz let me know if you think this is more or less readable (I believe fewer lines and variables the better, but interpretations on readability can be different)